### PR TITLE
Small performance fixes for getQueryDetails()

### PR DESCRIPTION
**Pull request by labkey-matthewb**
#### Rationale
Changes inspired by profiling getQueryDetails() as used by SM

#### Changes
cache _SqlParser objects, spending much more time creating these objects than actually parsing.  It would seem that you're supposed to hold on to them.

fixes to avoid calling new DefaultSchema() when we should be using existing object.

### CODE REVIEW

File: api/src/org/labkey/api/exp/query/SamplesSchema.java [hunk 1]
```diff
@@ -76,11 +76,17 @@ public boolean isAvailable(DefaultSchema schema, Module module)
 
             public QuerySchema createSchema(DefaultSchema schema, Module module)
             {
-                return new SamplesSchema(schema.getUser(), schema.getContainer());
+                return new SamplesSchema(schema);
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/exp/query/SamplesSchema.java:78]: It looks like we call `.setDefaultSchema()` when resolving schemas in `DefaultSchema.getSchema()`.  Does that not work in this case?  Does it happen too late?

File: query/src/org/labkey/query/sql/SqlParser.java [hunk 2]
```diff
@@ -1347,18 +1398,30 @@ public int LA(int i)
...
+        {
+            super(null);
             setTreeAdaptor(new LabKeyTreeAdaptor());
-            _errors = errors;
+            _errors = null;
             assert MemTracker.getInstance().put(this);
         }
 
+        public _SqlParser(String str, ArrayList<Exception> errors)
+		{
```
Comment (labkey-kevink) [inline query/src/org/labkey/query/sql/SqlParser.java]: weird indentation

File: query/src/org/labkey/query/sql/SqlParser.java [hunk 3]
```diff
@@ -116,29 +152,44 @@ public SqlParser(SqlDialect d, Container c)
         _container = c;
     }
 
+    /** see also _SqlParser.close() for AutoCloseable behavior */
+    private _SqlParser getAntlrParser()
+    {
+        synchronized (_parserPool)
+        {
+            _SqlParser ret = _parserPool.get();
```
Comment (labkey-kevink) [inline query/src/org/labkey/query/sql/SqlParser.java]: Does this block need to lock on _parserPool if `SoftPool.get` is synchronized?

**Review by labkey-kevink — CHANGES_REQUESTED** [review]

File: query/src/org/labkey/query/sql/SqlParser.java (hunk 3 above)
Comment (labkey-matthewb) [reply query/src/org/labkey/query/sql/SqlParser.java]: thanks, leftover from earlier edit.

File: query/src/org/labkey/query/sql/SqlParser.java (hunk 2 above)
Comment (labkey-matthewb) [reply query/src/org/labkey/query/sql/SqlParser.java]: Weird, intellij left in the tabs.

File: api/src/org/labkey/api/exp/query/SamplesSchema.java (hunk 1 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/exp/query/SamplesSchema.java:78]: The problem isn't really here in createSchema().  In other places people call "new SampleSchema()" directly, so I wanted to make a SampleSchema(schema) constructor.  Should probably go convert those places in this PR.

**Pull request closed without merging**